From: Steve Langasek <steve.langasek@canonical.com>
Date: Tue, 2 Jan 2024 10:23:29 +0100
Subject: support initramfsless boot

Bug-Ubuntu: https://bugs.launchpad.net/bugs/981314

Don't fail if --attach-to-session is passed and /dev/pts isn't mounted
yet, as this breaks booting without an initramfs.
---
 src/main.c | 11 +----------
 1 file changed, 1 insertion(+), 10 deletions(-)

diff --git a/src/main.c b/src/main.c
index 279e54a..e2f6f86 100644
--- a/src/main.c
+++ b/src/main.c
@@ -1900,6 +1900,7 @@ attach_to_running_session (state_t *state)
                 session = ply_terminal_session_new (NULL);
 
                 ply_terminal_session_attach_to_event_loop (session, state->loop);
+                state->session = session;
         } else {
                 session = state->session;
                 ply_trace ("session already created");
@@ -1911,12 +1912,6 @@ attach_to_running_session (state_t *state)
                                           (ply_terminal_session_hangup_handler_t)
                                           (should_be_redirected ? on_session_hangup : NULL),
                                           -1, state)) {
-                ply_save_errno ();
-                ply_terminal_session_free (session);
-                ply_buffer_free (state->boot_buffer);
-                state->boot_buffer = NULL;
-                ply_restore_errno ();
-
                 state->is_redirected = false;
                 state->is_attached = false;
                 return false;
@@ -1940,7 +1935,6 @@ attach_to_running_session (state_t *state)
 
         state->is_redirected = should_be_redirected;
         state->is_attached = true;
-        state->session = session;
 
         return true;
 }
@@ -2455,9 +2449,6 @@ main (int    argc,
                 state.should_be_attached = attach_to_session;
                 if (!attach_to_running_session (&state)) {
                         ply_trace ("could not redirect console session: %m");
-                        if (!no_daemon)
-                                ply_detach_daemon (daemon_handle, EX_UNAVAILABLE);
-                        return EX_UNAVAILABLE;
                 }
         }
 
